packages/endo/DESIGN.md

```diff
@@ -0,0 +1,287 @@
+> }
+> ```
+>
+> In that package, Endo would introduce the worker as a global.
+>
+> ```js
+> const worker = new MyPromiseWorker();
+> ```
+
+
+## Compartment Maps
+
+Endo at the command line works by generating a _compartment map_ from
+your application workspace and all of the `node_modules` it needs.
+A compartment map is similar to a lock file because it collects information
+from all of the installed modules.
+A compartment map describes how to construct compartments for each
+package in your application and link their module namespaces.
+
+> TODO
+>
+> For browser applications, a future version of Endo will ship
+> a `compartmap` tool that can be used as a `postinall` script
+> to generate a compartment map applications can use _during development_
+> to run on the client side without a build step.
+>
+> This compartment map will include the `browser` tag and all corresponding
+> module aliases from the `exports` field in `package.json`.
+> The Endo command line tool will always generate a compartment map
+> of its own without the `browser` tag and will never use the `compartmap.json`
+> file on disk.
+>
+> A compartment map is similar in spirit to an import map or a package lock.
+> The object fully describes how to compose a compartment DAG to run an
+> application.
+> A compartment map is intended to be generated only when dependencies or
+> compartment policies change, not to be drafted by hand, and certainly not to be
+> both manually and automatically maintained.
+>
+> As with an importmap, we can add a postinstall hook to package.json that will
+> generate a new compartmap each time the dependency graph of an application
+> changes.
+>
+> ```json
+> {
+>   "scripts": {
+>     "postinstall": "compartmap"
+>   }
+> }
+> ```
+>
+> We can include this in the scaffolding for compartmentalized applications.
+>
+> The `compartmap` tool generates a `compartmap.json` file, merging what
+> it finds in a `package-lock.json`, `yarn.lock` or some such shrinkwrap,
+> or just crawls `node_modules` and finds all the `package.json` files
+> as an input.
+> Then, it merges a `compolicy.json` if it finds one.
+> The default compartment policy is empty, which implies:
+>
+> * endowments only pass to the main compartment.
+> * no package has access to any built-in modules.
+
+The compartment map shape:
+
+```ts
+// CompartmentMap describes how to prepare compartments
+// to run an application.
+type CompartmentMap = {
+  tags: Tags,
+  main: CompartmentName,
+  compartments: Object<CompartmentName, Compartment>,
+  realms: Object<RealmName, Realm>, // TODO
+};
+
+// Tags are the build tags for the compartment.
+// These may include terms like "browser", meaning
+// each compartment uses the implementation of each
+// module suitable for use in a browser environment.
+type Tags = Array<Tag>;
+type Tag = string;
+
+// CompartmentName is an arbitrary string to name
+// a compartment for purposes of inter-compartment linkage.
+type CompartmentName = string;
```

Comment:
It sounds like, as far as the loader is concerned, the `CompartmentName` could just be an integer.. it's only purpose is to join two other tables together. But for debugging/readability purposes, I'm guessing you'd use (package name) + (version number)?

Reply:
Yes. `importPath` names the compartments according to the path relative to the entry compartment. Archives use name, version, and a counter (to disambiguate duplicates since I didn’t elect to puzzle out a consistent-coin-toss in the rare case that `npm` or `yarn` is forced to bury multiple copies of the same package name + version combination.

🤔 Actually, since `yarn` uses symlinks, I can eliminate the possibility of duplicates by canonicalizing the path of each package, which would require passing more of `fs` and wouldn’t entirely eliminate the possibility since we don’t require `yarn` to be used.

Reply:
If you *do* end up doing that canonicalization, you might want to follow it by a re-relativization step, to hide the developer's filesystem structure (e.g. I'd rather not publish archives to the world that embed `/home/warner/` in them). Basically the output should be a deterministic function of the source tree's contents (plus node_modules stuff), and moving the source tree elsewhere shouldn't cause the archive contents to change.

Reply:
Aye, you will see later in the stack, in `archive.js`, the bulk of the work is renaming the compartments and compartment links to launder the file system structure.